### DISCUSSION TITLE
internal viewer improvements

### DISCUSSION THREAD

**Discussion by mikekazakov** [mikekazakov/nimble-commander#251] [Ideas]
Issues:
* there should be a way to search backwards, e.g. shift return or opt-shift-g
* search does not work in PDFs; it would be nice if it did or if the search window was disabled when it doesn't work
* you cannot select text in PDFs for copying
* the percentage through indicator does not work for PDFs or other things in Preview mode
* there is no way to set the default view mode, I'd like to set it to Preview not Hex

**Comment by yougotwill** [comment]
@mikekazakov I've started using nimble commander as my daily explorer now that I'm more comfortable with it. Being able to set the internal viewer mode would be really useful. I hope this issue can get a 2nd look 🙏

**Reply by mikekazakov** [reply]
I'm not sure this can fit into v1.7, so `Alt+F3 -> Cmd+3` is the best approach for now.
If somebody wants to work on it - welcome.

**Comment by pretzelmonster** [comment]
Hi, jumping on your own very old post:
I have been using Nimble Commander daily for a long time, and the only thing that really also bugs me is the missing Copy functionality from the viewer, especially in PDFs. Search would certainly also be nice.
Any chance to get this implemented in the next 8 years? ;-)

Just to be sure not to forget: You have an excellent tool here. Thanks for the tremendeous job - and the idea to put it out as Open Source now!

Best regards!

**Reply by mikekazakov** [reply]
Hey, thanks for your kind words!
Now that the project is open source, any chance you, @pretzelmonster, might contribute and help move things forward? 😉
Unfortunately, there’s no magic recipe - Nimble Commander currently has a single developer and maintainer. If more people want to see features like this implemented faster, getting involved and contributing would make a big difference!

**Reply by pretzelmonster** [reply]
Hi,
I'd indeed be interested to help, but I can't really program, sorry. While I can use Unity to create an app, this is beyond my skillls. On the other hand I doubt it is really difficult to at least implement the copy feature. Text can already be marked, so I wonder how difficult it can be to poll for "command c" and copy the marked text to the clipboard? I can understand that a find feature might be not as easy to implement.

Where I can really offer my help would be to get the app translated to other languages. I have helped with this purely text-based support for a few other open source projects, in case you plan that for the future?
Please contact me in case you would like to go in that direction, but I'm not sure how flexible Nimble Commander has been set up regarding multiple languages.

**Reply by mikekazakov** [reply]
> On the other hand I doubt it is really difficult to at least implement the copy feature. Text can already be marked, so I wonder how difficult it can be to poll for "command c" and copy the marked text to the clipboard? I can understand that a find feature might be not as easy to implement.

Well, to begin with, the actual implementation of the QuickLook preview runs in a separate process (you can observe "QuickLookUIService"), which is mirrored into the app via internal MacOS magic, completely hidden from any control by the calling application - Nimble Commander in this case. The shallow shim (`QLPreviewView`), which applications are supposed to communicate with, doesn't accept neither regular hotkeys nor normal commands, like "copy:". If you know any application with available source which implements what you want from QuickLook - I'm very open to looking at that implementation. So far I'm not sure it's even implementable at all, there's literally an entry in the [F.A.Q.](https://github.com/mikekazakov/nimble-commander/blob/main/Docs/Help.md#frequently-asked-questions) about this .

> Where I can really offer my help would be to get the app translated to other languages. I have helped with this purely text-based support for a few other open source projects, in case you plan that for the future?
Please contact me in case you would like to go in that direction, but I'm not sure how flexible Nimble Commander has been set up regarding multiple languages.

Nimble Commander can be translated to other languages without touching a single line of code, that's been like this from the beginning. The process of translation though requires learning about how localizations generally work on the Apple platforms. In v1.8 I aim to migrate the string resources to the new Strings Catalog format, which should make the process easier and might even provide a decent experience directly in Xcode.
